### PR TITLE
Http server

### PR DESCRIPTION
**Pull request by dkoshkin**
Fixes https://github.com/apprenda/kismatic/issues/878

Added a new `server` package that will contain all code for the KET API.
* Using [httprouter](https://github.com/julienschmidt/httprouter) - a lightweight router
* Using [negroni](https://github.com/urfave/negroni) - some important middleware(Logger, Recovery)
* Added a `/heatlhz` route
* Added a `main()` for simpler/faster testing cycle

Added the new `server` command

TODO:
- [ ] Regenerate CLI docs

### CODE REVIEW

File: pkg/server/http/server_test.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+func TestHandler(t *testing.T) {
+	if testing.Short() {
+		return
+	}
+	logger := DefaultLogger(os.Stdout, "[tester] ")
+	s, err := New(logger, "", "", "", -1, -1)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	ts := httptest.NewServer(s.httpServer.Handler)
```
Comment (alexbrand) [inline pkg/server/http/server_test.go]: This seems like a redundant test to me, as we are not testing the actual server, and we already have a test for the handler.

File: pkg/server/http/server.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/urfave/negroni"
+)
+
+const (
+	writeTimeout = 10 * time.Second
+	readTimeout  = 10 * time.Second
+)
+
+// Server holds the details for an http(s) server
+// Leave cert and key empty to use a self signed CA
```
Comment (alexbrand) [inline pkg/server/http/server.go]: I don't see a `cert` or `key` field in the Server struct. We can probably remove from this comment?

If the Server struct needs to be initialized via `New`, I think I'd make it unexported.

File: pkg/server/http/server.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	logger     *log.Logger
+}
+
+const (
+	loggerPrefix   = "[kismatic] "
+	defaultPort    = "8443"
+	defaultTimeout = 15 * time.Second
+)
+
+// New creates a configured http server with a gorilla mux router
```
Comment (alexbrand) [inline pkg/server/http/server.go]: I think we can remove implementation details (gorilla, negroni) out of this comment, as they are already outdated.

File: pkg/server/http/server.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+// And negroni Recovery and Logger middleware
+// If certificates are not provided, a self signed CA will be used
+// Use negative numbers for default read and write timeouts
```
Comment (alexbrand) [inline pkg/server/http/server.go]: Not a fan of this. I am wondering if we are better off not having defaults.

File: pkg/server/http/server.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+		},
+		logger: logger,
+	}
+
+	return s, nil
+}
+
+// RunTLS support
+func (s *Server) RunTLS() error {
+	s.logger.Printf("Listening on 0.0.0.0%s\n", s.httpServer.Addr)
```
Comment (alexbrand) [inline pkg/server/http/server.go]: Typo `0.0.0.0 + s.httpServer.Addr`?

File: pkg/tls/ca.go [hunk 6]
```diff
@@ -35,12 +36,15 @@ func NewCACert(csrFile string, commonName string, expiry string) (key, cert []by
 	if err != nil {
 		return nil, nil, fmt.Errorf("error opening %q", csrFile)
 	}
+	return NewCACertFromIOReader(f, commonName, expiry)
+}
+
+func NewCACertFromIOReader(r io.Reader, commonName string, expiry string) (key, cert []byte, err error) {
```
Comment (alexbrand) [inline pkg/tls/ca.go]: Thoughts on dropping IO from the func's name here and renaming `r` to `csrReader`?

File: pkg/server/http/server.go (hunk 5 above)
Comment (dkoshkin) [reply pkg/server/http/server.go]: `s.httpServer.Addr` will be `:443`, did you mean the missing `:`?

File: pkg/server/http/server.go (hunk 2 above)
Comment (alexbrand) [reply pkg/server/http/server.go]: Actually, not sure about making Server unexported.

File: pkg/server/http/server.go (hunk 5 above)
Comment (alexbrand) [reply pkg/server/http/server.go]: Oh I see. So there's no way to start the server on a specific interface?

**Comment by alexbrand** [comment]
This looks good to me overall. The only concern I have at this point is the license of `github.com/julienschmidt/httprouter`. Seems to be a non-standard license? I can't tell.

**Comment by dkoshkin** [comment]
Merging for now, we can discuss the license later on.

**Pull request merged**